### ISSUE TITLE
[GJEP] NamedArgs for generative functions

### ISSUE THREAD

**Issue by mirkoklukas** [genjax-community/genjax#918]
I found it useful to have something like a named tuple object as argument for my generative functions. Here's what I came up with:

```
from builtins import property as _property, tuple as _tuple
from jax.tree_util import register_pytree_node_class


@register_pytree_node_class
class NamedArgs(tuple):
    def __new__(cls, *args, **kwargs):
        return _tuple.__new__(cls, list(args) + list(kwargs.values()))
    
    def __init__(self, *args, **kwargs):
        self._d = dict()
        for k,v in kwargs.items():
            self._d[k] = v
            setattr(self, k, v)

    def __getitem__(self, k: str):
        return self._d[k]

    def tree_flatten(self):
        return (self, None)

    @classmethod
    def tree_unflatten(cls, aux_data, children):
        return cls(*children)
```

**Comment by femtomc** [comment]
Thanks @mirkoklukas, this is quite nice. I may put this into the `Pytree` general class, as a common thing that people can use and build with.

**Comment by littleredcomputer** [comment]
Why not just use a dict? It is the Pythonic way to associate names to values. I understand that special handling is needed for JAX tracing and that the tuple is currently baked deeply into the argument structure for GFs, but from the point of view of the UX, it might be preferable to let the user write `{"x": v}` rather than `NamedArgs(x=v)`. Convert it to whatever works on the inside

**Comment by mirkoklukas** [comment]
It's more of a hack but keeps the argument order and I can also use `NamedArgs(1,2,3,z=4)` like in a func signature. The best thing would be to have gen functions take keyword arguments I guess.

**Comment by femtomc** [comment]
@littleredcomputer How would the calling convention work for generative functions? They expect tuples.

**Comment by littleredcomputer** [comment]
```
@genjax.static
def f(d):
    return genjax.normal(d["mean"], d["stdev"]) @ "y"

print(f.simulate(key, ({
    "mean": 10.0,
    "stdev": 0.1
},)))


@genjax.static
def g(*, mean, stdev):
    return genjax.normal(mean, stdev) @ "y"

print(g.simulate(key, (), **{
    "mean": 10.0,
    "stdev": 0.1
}))
```

The first one works but it's clunky because you need to wrap the dict in a 1-tuple. The second doesn't because simulate won't propagate kwargs.

It would be nice to place no requirements on the shape of argument lists of functions that the user writes herself (so that if she wants to enforce a keyword only style, she may, by using Python's existing feature for this).  I wasn't aware (viscerally anyway) that GFs could not take keyword args.

I wish there was some other way to pass the key in, so that simulate (for example) would not have to tuple the arglist

**Comment by sritchie** [comment]
This is now solved in the upcoming 0.4.0 release with the ability to use keyword args with generative functions; we'll get this documented in the next cycle. Closing this for now, keep an eye out for the shipped email.
